fix: avoid showing stale diagnostics in handlers

faergeek opened the pull request:
Fixes #35884

Problem: Since `once_buf_loaded` might call a callback passed to it at a later point (which is it's reason to exist) that callback might end up referring to stale diagnostics in a handler's `show` function. For example, if we first call `vim.diagnostic.set` for an unloaded buffer, then call `vim.diagnostic.reset` and only then load the buffer, we might still see diagnostics from `vim.diagnostic.set` call, which are stale at this point.

Solution: only keep one autocommand from the most reset `show` call and delete it when `hide` is called.

justinmk commented:
This looks fragile, because 4 different places need to manage things carefully. 

Can we at least deduplicate the pattern (use a local function). And maybe choose names that are less verbose.

faergeek committed:
fix: avoid showing stale diagnostics in handlers

Fixes #35884

Problem: Since `once_buf_loaded` might call a callback passed to it at a
later point (which is it's reason to exist) that callback might end up
referring to stale diagnostics in a handler's `show` function. For
example, if we first call `vim.diagnostic.set` for an unloaded buffer,
then call `vim.diagnostic.reset` and only then load the buffer, we might
still see diagnostics from `vim.diagnostic.set` call, which are stale at
this point.

Solution: only keep one autocommand from the most reset `show` call and
delete it when `hide` is called.
